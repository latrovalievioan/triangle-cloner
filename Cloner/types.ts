export type BaseElementProps = {
  id?: string;
  key?: string
}

export type TextElement = {
  type: "TEXT_ELEMENT",
  props: {
    nodeValue: string,
    children: never[]
  }
}

export type ClonerElement<PropsType extends BaseElementProps = BaseElementProps> = {
  type: string;
  props: PropsType & {
    children: (ClonerElement | TextElement)[]
  }
}
