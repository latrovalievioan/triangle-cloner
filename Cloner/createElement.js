// const element = (
//   <div id="foo">
//     <a>bar</a>
//     <b />
//   </div>
// )
// const container = document.getElementById("root")
// ReactDOM.render(element, container)
//
// IS ACTUALLY --------xxxxxxxx---------xxxxxxxx----------xxxxxxxx---------xxxxxx---------
//
// const element = React.createElement(
//   "div",
//   { id: "foo" },
//   React.createElement("a", null, "bar"),
//   React.createElement("b")
// )
// const container = document.getElementById("root")
// ReactDOM.render(element, container)

export const createElement = (type, props, ...children) => {
  return ({
    type,
    props: {
      ...props,
      children,
    }
  })
}

export const createTextElement = (text) => {
  return ({
    type: "TEXT_ELEMENT",
    props: {
      nodeValue: text,
      children: [],
    }
  })
}
