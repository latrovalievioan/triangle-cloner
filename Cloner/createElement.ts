type Props = {
  [k: string]: any
}

type Children = (string | ClonerElement)[]

export type ClonerElement = {
  type: string,
  props: Partial<Props & Children>
}

export const createElement = (
  type: string,
  props: Props,
  ...children: Children
): ClonerElement => ({
    type,
    props: {
      ...props,
      children: children.map(c => 
        typeof c === 'string'
          ? createTextElement(c)
          : c
      ),
    }
})


const createTextElement = (text: string) => ({
  type: "TEXT_ELEMENT",
  props: {
    nodeValue: text,
    children: [],
  }
})
