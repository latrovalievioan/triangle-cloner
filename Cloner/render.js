export const render = (element, container) => {
  const dom = element.type === "TEXT_ELEMENT" 
    ? document.createTextNode("") 
    : document.createElement(element.type)

  Object.keys(element.props).forEach(k => {
    if (k === 'children') return
    dom[k] = element.props[k]
  })

  element.props.children.forEach(c => {
    render(c, dom)
  })

  container.appendChild(dom)
}
