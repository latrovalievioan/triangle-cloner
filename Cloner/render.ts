import { ClonerElement, TextElement } from './types';

const isHTMLElement = (element: HTMLElement | Text): element is HTMLElement => {
  return element.nodeName !== '#text';
};

export const render = (
  element: ClonerElement | TextElement,
  container: HTMLElement
): void => {
  const dom =
    element.type === 'TEXT_ELEMENT'
      ? document.createTextNode('')
      : document.createElement(element.type);

  Object.keys(element.props).forEach((k) => {
    if (k === 'children') return;

    dom[k] = element.props[k];
  });

  if (isHTMLElement(dom)) {
    element.props.children.forEach((c) => render(c, dom));
  }

  container.appendChild(dom);
};
