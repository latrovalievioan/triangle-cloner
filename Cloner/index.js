import { createElement } from "./createElement.js";
import { render } from "./render.js"



export const Cloner = {
  createElement,
  render
}
