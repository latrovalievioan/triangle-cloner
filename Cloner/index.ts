import { createElement } from './createElement.ts';
import { render } from './render.ts';

export default {
  createElement,
  render,
};
