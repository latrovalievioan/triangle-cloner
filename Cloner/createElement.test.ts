import Cloner from './index';
import { describe, test, expect } from 'vitest';

describe('createElement', () => {
  test("doesn't crash", () => {
    expect(() => Cloner.createElement('div', { id: 'div' })).not.toThrowError();
  });

  test('element has the correct type', () => {
    const el = Cloner.createElement('div', { id: 'div' });
    expect(el.type).toEqual('div');
  });

  test('element has the correct props', () => {
    const props = {
      id: 'div',
      name: 'dummy_div',
      count: 0,
    };
    const el = Cloner.createElement('div', props);

    expect(el.props).toMatchObject(props);
  });

  test('element has the correct children count', () => {
    const el = Cloner.createElement('div', {});
    const e2 = Cloner.createElement('div', {});
    const e3 = Cloner.createElement('div', {});
    const e4 = Cloner.createElement('div', {}, 'hehe', el, e2, e3);

    expect(e4.props.children).toHaveLength(4);
  });

  test('creates textElement from string children', () => {
    const el = Cloner.createElement('h1', {}, 'hehe');
    const textEl = {
      type: 'TEXT_ELEMENT',
      props: {
        nodeValue: 'hehe',
        children: [],
      },
    };
    expect(el.props.children[0]).toStrictEqual(textEl);
  });
});
