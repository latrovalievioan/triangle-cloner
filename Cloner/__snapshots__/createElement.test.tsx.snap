// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`createElement > Snapshot 1`] = `
{
  "props": {
    "children": [
      {
        "props": {
          "children": [
            {
              "props": {
                "children": [
                  {
                    "props": {
                      "children": [],
                      "nodeValue": "1",
                    },
                    "type": "TEXT_ELEMENT",
                  },
                ],
              },
              "type": "li",
            },
            {
              "props": {
                "children": [
                  {
                    "props": {
                      "children": [],
                      "nodeValue": "2",
                    },
                    "type": "TEXT_ELEMENT",
                  },
                ],
              },
              "type": "li",
            },
            {
              "props": {
                "children": [
                  {
                    "props": {
                      "children": [],
                      "nodeValue": "3",
                    },
                    "type": "TEXT_ELEMENT",
                  },
                ],
              },
              "type": "li",
            },
            {
              "props": {
                "children": [
                  {
                    "props": {
                      "children": [],
                      "nodeValue": "4",
                    },
                    "type": "TEXT_ELEMENT",
                  },
                ],
              },
              "type": "li",
            },
            {
              "props": {
                "children": [
                  {
                    "props": {
                      "children": [],
                      "nodeValue": "5",
                    },
                    "type": "TEXT_ELEMENT",
                  },
                ],
              },
              "type": "li",
            },
          ],
        },
        "type": "ul",
      },
    ],
  },
  "type": "div",
}
`;

exports[`createElement > ne6to si 1`] = `
{
  "props": {
    "children": [
      {
        "props": {
          "children": [
            {
              "props": {
                "children": [
                  {
                    "props": {
                      "children": [],
                      "nodeValue": "1",
                    },
                    "type": "TEXT_ELEMENT",
                  },
                ],
              },
              "type": "li",
            },
            {
              "props": {
                "children": [
                  {
                    "props": {
                      "children": [],
                      "nodeValue": "2",
                    },
                    "type": "TEXT_ELEMENT",
                  },
                ],
              },
              "type": "li",
            },
            {
              "props": {
                "children": [
                  {
                    "props": {
                      "children": [],
                      "nodeValue": "3",
                    },
                    "type": "TEXT_ELEMENT",
                  },
                ],
              },
              "type": "li",
            },
            {
              "props": {
                "children": [
                  {
                    "props": {
                      "children": [],
                      "nodeValue": "4",
                    },
                    "type": "TEXT_ELEMENT",
                  },
                ],
              },
              "type": "li",
            },
            {
              "props": {
                "children": [
                  {
                    "props": {
                      "children": [],
                      "nodeValue": "5",
                    },
                    "type": "TEXT_ELEMENT",
                  },
                ],
              },
              "type": "li",
            },
          ],
        },
        "type": "ul",
      },
    ],
  },
  "type": "div",
}
`;
